src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -3756,6 +3777,34 @@ ErrorIfQueryHasUnroutableModifyingCTE(Query *queryTree)
 }
 
 
+#if PG_VERSION_NUM >= PG_VERSION_14
+
+/*
+ * ErrorIfQueryHasCTEWithSearchClause checks if the query contains any common table
+ * expressions with search clause and errors out if it does.
+ */
+static DeferredErrorMessage *
+ErrorIfQueryHasCTEWithSearchClause(Query *queryTree)
+{
+	CommonTableExpr *cte = NULL;
+	foreach_ptr(cte, queryTree->cteList)
```

Comment:
This check does not cover CTEs that are not in the top level:
```SQL


SELECT * FROM (
WITH RECURSIVE search_graph(f, t, label) AS
  (SELECT *
   FROM graph0 g
   WHERE f = 1 
   UNION ALL SELECT g.*
   FROM graph0 g,
        search_graph sg
   WHERE g.f = sg.t
     AND g.f = 1  ) SEARCH DEPTH FIRST BY f, t
SET seq
SELECT *
FROM search_graph
ORDER BY seq) as foo;
ERROR:  column reference "seq" is ambiguous
CONTEXT:  while executing command on localhost:9701

SELECT * FROM (
WITH RECURSIVE search_graph(f, t, label) AS
  (SELECT *
   FROM graph0 g
   WHERE f = 1 
   UNION ALL SELECT g.*
   FROM graph0 g,
        search_graph sg
   WHERE g.f = sg.t
     AND g.f = 1  ) SEARCH DEPTH FIRST BY f, t
SET seq
SELECT *
FROM search_graph
ORDER BY seq) as foo;
ERROR:  column reference "seq" is ambiguous
CONTEXT:  while executing command on localhost:9701
```

I think we need to traverse the query tree. But, we should only do if `hasRecursive` is true, to keep the overhead only applicable to queries with recursive ctes

Reply:
I think `hasRecursive` is only valid for the current `Query`. It doesn't reflect all of the subqueries, so I couldn't use it.